src/openms/source/ANALYSIS/TOPDOWN/DeconvolutedSpectrum.cpp

```diff
@@ -0,0 +1,654 @@
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Kyowon Jeong, Jihyung Kim $
+// $Authors: Kyowon Jeong, Jihyung Kim $
+// --------------------------------------------------------------------------
+
+#include "include/OpenMS/ANALYSIS/TOPDOWN/DeconvolutedSpectrum.h"
+
+namespace OpenMS
+{
+  DeconvolutedSpectrum::DeconvolutedSpectrum()
+  {
+  }
+
+  DeconvolutedSpectrum::DeconvolutedSpectrum(MSSpectrum &s, int n) :
+      spec(&s), scanNumber(n)
+  {
+  }
+
+  DeconvolutedSpectrum::~DeconvolutedSpectrum()
+  {
+    std::vector<PeakGroup>().swap(peakGroups);
+  }
+
+  bool DeconvolutedSpectrum::empty() const
+  {
+    return peakGroups.empty();
+  }
+
+  MSSpectrum DeconvolutedSpectrum::toSpectrum()
+  {
+    auto outSpec = MSSpectrum(*spec);
+    outSpec.clear(false);
+    for (auto &pg : peakGroups)
+    {
+      if (pg.peaks.empty())
+      {
+        continue;
+      }
+      auto p = Peak1D(pg.monoisotopicMass, pg.intensity);
+      outSpec.push_back(p);
+    }
+    if (precursorPeakGroup != nullptr && !spec->getPrecursors().empty())
+    {
+      std::vector<Precursor> precursors;
+      Precursor precursor(spec->getPrecursors()[0]);
+      precursor.setCharge(precursorPeakGroup->maxQScoreCharge);
+      precursor.setMZ(precursorPeakGroup->monoisotopicMass);
+      precursor.setIntensity(precursorPeakGroup->intensity);
+      precursors.push_back(precursor);
```

Comment:
might be faster to provide a mutable getter to Precursors() and then use outSpec.getPrecursors().emplace_back({precursor}) if this part is speed relevant